src/Common/src/TypeSystem/IL/Stubs/Marshal/Marshaller.cs

```diff
@@ -0,0 +1,484 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using Internal.TypeSystem;
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.IL.Stubs.Marshal
+{
+    enum MarshallerKind
+    {
+        Unknown,                // Default marshaller and for types that PrefastImporter doesn't understand
+        BlittableValue,
+        BlittableArray,
+        BlittableByRef,
+        String,
+        Bool,
+        StringBuilder,
+        SafeHandle,
+        VoidReturn
+    }
+    public enum MarshalAttributes
+    {
+        None = 0,
+        In = 1,
+        Out = 2,
+        Return = 4,
+        Optional = 8
+    };
+
+    // Each type of marshaller knows how to generate the marshalling code for the argument it marshals.
+    // Marshallers contain method related marshalling information (which is common to all the Marshallers)
+    // and also argument specific marshalling informaiton
+    abstract class Marshaller
+    {
+        public PInvokeMethodData PInvokeMethodData;
+
+        #region Instance state information
+        protected int ArgumentIndex;
+        public MarshalAttributes MarshalAttributes;
+        public MarshallerKind MarshallerKind;
+        public TypeDesc NativeType;
+        public TypeDesc ManagedType;
+        #endregion
+
+        #region Code Stream Forwarders
+        public ILEmitter ILEmitter { get { return PInvokeMethodData.ILEmitter; } }
```

Comment:
This does not look right. EmitMarshallingIL should get the emitters as argument, they should not be smuggled it via PInvokeMethodData.